ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceCall.java

```diff
@@ -0,0 +1,251 @@
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.google.common.base.Suppliers;
+import org.apache.tuweni.bytes.Bytes;
+
+public class TraceCall implements JsonRpcMethod {
+  private static final String TRACE_NODE_PROPERTY_NAME = "trace";
+  private static final String OUTPUT_NODE_PROPERTY_NAME = "output";
+  private static final String STATE_DIFF_NODE_PROPERTY_NAME = "stateDiff";
+  private static final String VM_TRACE_NODE_PROPERTY_NAME = "vmTrace";
+
+  private final Supplier<BlockchainQueries> blockchainQueries;
+  private final ProtocolSchedule protocolSchedule;
+  private final TransactionSimulator transactionSimulator;
+
+  public TraceCall(
+      final BlockchainQueries blockchainQueries,
+      final ProtocolSchedule protocolSchedule,
+      final TransactionSimulator transactionSimulator) {
+    this.blockchainQueries = Suppliers.ofInstance(blockchainQueries);
+    this.protocolSchedule = protocolSchedule;
+    this.transactionSimulator = transactionSimulator;
+  }
+
+  @Override
+  public String getName() {
+    return transactionSimulator != null ? RpcMethod.TRACE_CALL.getMethodName() : null;
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    final TraceTypeParameter traceTypeParameter =
+        requestContext.getRequiredParameter(1, TraceTypeParameter.class);
+    final Optional<BlockParameter> maybeBlockParameter =
+        requestContext.getOptionalParameter(2, BlockParameter.class);
+
+    final Optional<BlockHeader> maybeBlockHeader = resolveBlockHeader(maybeBlockParameter);
+
+    if (maybeBlockHeader.isEmpty()) {
+      throw new IllegalStateException("Invalid block parameter.");
+    }
+
+    final Set<TraceTypeParameter.TraceType> traceTypes = traceTypeParameter.getTraceTypes();
+    final DebugOperationTracer tracer = new DebugOperationTracer(buildTraceOptions(traceTypes));
+    final Optional<TransactionSimulatorResult> maybeSimulatorResult =
+        transactionSimulator.process(
+            JsonCallParameterUtil.validateAndGetCallParams(requestContext),
+            buildTransactionValidationParams(),
+            tracer,
+            maybeBlockHeader.get());
+
+    if (maybeSimulatorResult.isEmpty()) {
+      throw new IllegalStateException("Invalid transaction simulator result.");
+    }
+
+    final ObjectMapper mapper = new ObjectMapper();
+    final ObjectNode resultNode = mapper.createObjectNode();
+
+    resultNode.put(OUTPUT_NODE_PROPERTY_NAME, maybeSimulatorResult.get().getOutput().toString());
+
```

Comment:
is there a reason we have to do direct json construction with a mapper instead of using a `org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.trace.*Result` class to wrap the result and let the framework do the json mapping?

Reply:
Good advice. I've added a result class and it made the reduced the complexity quite a lot.